:books: [#42] add example use to docs (#55)
maykinmedia/sharing-configs: README.rst

            the JSON generated by the above function.
            """              
            decoded_content = content.decode("utf-8")
            theme_dict = json.loads(decoded_content)        
            return ColorTheme.objects.create(**theme_dict)       

That takes care of the import and export functionality for exchaning color-themes. To make it actually working, 
we complete this example with some additional code.
Create a ``context_processors.py`` file, to pass the currently configured theme to the template context:

.. code-block:: python

    def set_admin_color(request:object)->dict:
        """
        create a dictionary of color variables to pass to the base_site.html Django admin page
        """
        conf = Configuration.get_solo()
        theme = conf.theme
        if theme is None:
            return {}

        return {
            "primary": theme.primary,
            "secondary": theme.secondary,            
            "primary_fg": theme.primary_fg,
            "accent": theme.accent
            
        }

Finally, pass the theme context variables to an overriden ``base_site.html`` in the templates folder.

.. code-block::

>   {% extends "admin/base_site.html" %}

Reviewer: This code block doesnt seem to render in markdown. Maybe indent is wrong or it needs a return after the block statement.

    :target: https://pypi.org/project/sharing_configs/

.. _Maykin Media B.V.: https://www.maykinmedia.nl
.. _Sharing Configs API: https://github.com/maykinmedia/sharing-configs-api.git

Example
=======

> Simple example for changing color theme in the django admin:

Reviewer: The example should focus on the sharing aspect. Now, the example focuses on the color switching. So, let's rewrite it a little bit:

```suggestion

We can use the Sharing Configs library to exchange color-themes for the Django admin with other users. For this, we need a model that stores the color-theme, and use the Sharing Configs mixins to import and export the color-theme.
```

Create two models: Configuration and Theme

.. code-block:: python

>     from django.db import models

Author: Added

    from django.utils.translation import gettext_lazy as _
    from solo.models import SingletonModel

    class Configuration(SingletonModel):
        """

>         configuration for export/import of color theme in the Django admin

Reviewer: ```suggestion
        Configuration that holds the current theme.
```

Author: yes

        """

>         theme = models.ForeignKey("Theme", on_delete=models.CASCADE, related_name="theme")

Author: Adjusted

        def __str__(self) -> str:

>             return self.theme.name

Author: Done

    class Theme(models.Model):
        """object to change color theme in Django admin"""

        name = models.CharField(_("theme_name"), max_length=120)
        primary = models.CharField(_("primary color"), max_length=60)
        secondary = models.CharField(_("secondary color"), max_length=60)
        accent = models.CharField(_("accent color"), max_length=60)
        primary_fg = models.CharField(_("primary_fg"), max_length=60)
        
        
Add two mixins provided by the sharing-configs app to the ThemeAdmin. 
Thus you will get an interface to export and import ColorTheme objects to and from the API.  

Customize two mixins methods: **get_sharing_configs_export_data** to convert model object to bytes and 
**get_sharing_configs_import_data** to convert bytes to an object.

> Register both models with the admin site.

Author: Done

.. code-block:: python

>     import json

Author: Added

    from django.contrib import admin
    from django.forms.models import model_to_dict
    from django.shortcuts import get_object_or_404

    from sharing_configs.admin import SharingConfigsExportMixin, SharingConfigsImportMixin
    from sharing_configs.forms import ExportToForm, ImportForm

    from .models import Configuration, Theme

    class ThemeAdmin(SharingConfigsExportMixin, SharingConfigsImportMixin, admin.ModelAdmin):        

        def get_sharing_configs_export_data(self, obj: object) -> bytes:

>             """return  django theme model object as a byte like object"""

Author: Done

            theme = get_object_or_404(Theme, id=obj.id)
            theme_dict = model_to_dict(theme)
            theme_dict.pop("id", None)
            dump_json_theme = json.dumps(cleaned_theme_dict, sort_keys=True, default=str)        
            return dump_json_theme.encode("utf-8")

        def get_sharing_configs_import_data(self, content: bytes) -> object:
            """
            convert byte string into a dictionary and create a color-theme object;           

>             example api response: b'{"accent": "#f8f8f8", "name": "some-name", "primary": "#ab8585", "primary_fg": "#1a2b3c", "secondary": "#315980"}'

Author: Done

            """              
            decoded_content = content.decode("utf-8")
            theme_dict = json.loads(decoded_content)        
            return ColorTheme.objects.create(**theme_dict)       

> Create a ``context_processors.py`` file.  
> Add ``yourapp.context_processors.set_color`` to the list of TEMPLATES context_processors in ``settings.py``

Reviewer: ```suggestion
That takes care of the import and export functionality for exchaning color-themes. To make it actually working, we complete this example with some additional code.

Create a ``context_processors.py`` file, to pass the currently configured theme to the template context:
```

Author: Done

.. code-block:: python

    def set_admin_color(request:object)->dict:
        """
        create a dictionary of color variables to pass to the base_site.html Django admin page
        """
        conf = Configuration.get_solo()

>         color_theme_obj = Theme.objects.filter(theme=conf).last()
>         primary_color = color_theme_obj.primary
>         secondary_color = color_theme_obj.secondary
>         accent = color_theme_obj.accent
>         primary_fg = color_theme_obj.primary_fg    
>         return {
>             "primary": primary_color,
>             "secondary": secondary_color,            
>             "primary_fg": primary_fg,
>             "accent": accent,
>             
>         }

Reviewer: Let's make this a little bit more compact. Also the FK-relation is weird to read like this:

```suggestion
        theme = conf.theme
        if theme is None:
            return {}
       
        return {
            "primary": theme.primary,
            "secondary": theme.secondary,            
            "primary_fg": theme.primary_fg,
            "accent": theme.accent,
        }
```

Author: Agree

> Pass created variables to the ``base_site.html`` in the templates.

Author: Done

.. code-block::
  {% extends "admin/base_site.html" %}
    {% block extrastyle %}
    <html
        style=" --primary:{{primary}}; --secondary:{{secondary}}; --accent:{{accent}}; --primary_fg:{{primary_fg}};">
    </html>
  {% endblock %}

> Now you can choose an available color theme via related congiguration object on its admin page.

Author: Changed

> Now you can choose an available color theme via related congiguration object on its admin page.

Author: Done
